File: demo-website/src/services/totpService.ts
/**
 * TOTP generation and verification, for the relying party.
 *
 * This is the other half of the handshake. `authenticator-app` computes a code
 * from a secret and the clock; this computes the same code independently and
 * compares. Neither side sends the other anything to make it happen.
 *
 * Implemented with the Web Crypto API (`crypto.subtle`), which is available in
 * every browser and in Node 18+, so the exact same module runs in the demo and
 * in a Cloud Function. See `docs/totp-working.md` for the algorithm.
 *
 * ## Read this before copying the pattern
 *
 * Verification here runs **in the browser**. That is wrong for production, and
 * the project says so loudly - a client can claim any result it likes, so a
 * browser saying "the code was valid" proves nothing to a server.
 *
 * The demo does it anyway because Cloud Functions require a billing plan, and a
 * student should not have to enter a credit card to see how MFA works. The code
 * is written so that moving it server-side is a copy-paste: no DOM, no React,
 * no Firebase imports. `examples/integration-examples/` shows it deployed
 * properly.
 */

/** RFC 4648 Base32 alphabet. Base32, not Base64 - see `docs/totp-working.md`. */
const BASE32_ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';

/** Seconds a code is valid. */
export const TOTP_PERIOD = 30;

/** Digits shown to the user. */
export const TOTP_DIGITS = 6;

/**
 * Time steps accepted either side of the current one.
 *
 * One step is about 90 seconds of total tolerance, which covers a slightly wrong
 * phone clock and a user who starts typing at second 29. Widening it is
 * tempting and wrong: every extra step multiplies the brute-force surface and
 * extends how long a stolen code stays usable.
 */
export const TOTP_DRIFT_WINDOW = 1;

export type TotpAlgorithm = 'SHA1' | 'SHA256' | 'SHA512';

/** Maps the `otpauth://` spelling onto the Web Crypto name. */
function subtleHashName(algorithm: TotpAlgorithm): string {
  switch (algorithm) {
    case 'SHA256':
      return 'SHA-256';
    case 'SHA512':
      return 'SHA-512';
    case 'SHA1':
      return 'SHA-1';
  }
}

/** Thrown when a secret is not decodable Base32. */
export class InvalidSecretError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'InvalidSecretError';
  }
}

/**
 * Decodes Base32 text into bytes.
 *
 * Tolerant of lowercase, spaces, hyphens and `=` padding, because those are all
 * ways the same key legitimately arrives from a user pasting it.
 */
export function base32Decode(input: string): Uint8Array {
  const cleaned = input
    .toUpperCase()
    .replace(/[\s-]/g, '')
    .replace(/=+$/, '');

  if (cleaned === '') {
    throw new InvalidSecretError('The secret is empty.');
  }

  const output: number[] = [];
  let buffer = 0;
  let bits = 0;

  for (const char of cleaned) {
    const value = BASE32_ALPHABET.indexOf(char);
    if (value < 0) {
      throw new InvalidSecretError(`"${char}" is not a valid Base32 character.`);
    }

    buffer = (buffer << 5) | value;
    bits += 5;

    if (bits >= 8) {
      output.push((buffer >> (bits - 8)) & 0xff);
      bits -= 8;
    }
  }

  if (output.length === 0) {
    throw new InvalidSecretError('The secret is too short to be a key.');
  }

  return new Uint8Array(output);
}

/** The time step for an instant: how many whole periods since the epoch. */
export function counterFor(at: Date = new Date(), period: number = TOTP_PERIOD): number {
  return Math.floor(at.getTime() / 1000 / period);
}

/**
 * Computes the code for a given counter.
 *
 * Async because Web Crypto is async. That propagates outward, which is why
 * every verification function in this file returns a promise.
 */
export async function generateForCounter(
  secret: string,
  counter: number,
  digits: number = TOTP_DIGITS,
  algorithm: TotpAlgorithm = 'SHA1',
): Promise<string> {
  const keyBytes = base32Decode(secret);

  // Step 1: the counter as an 8-byte big-endian integer.
  const message = new ArrayBuffer(8);
  // `setBigUint64` rather than two 32-bit halves: the counter exceeds 2^32 in
  // the year 6053, but using the 64-bit write keeps it correct by construction
  // rather than by luck.
  new DataView(message).setBigUint64(0, BigInt(counter), false);

  // Step 2: HMAC.
  const key = await crypto.subtle.importKey(
    'raw',
    keyBytes as unknown as BufferSource,
    { name: 'HMAC', hash: { name: subtleHashName(algorithm) } },
    false,
    ['sign'],
  );
  const mac = new Uint8Array(await crypto.subtle.sign('HMAC', key, message));

  // Step 3: dynamic truncation - the offset comes from the hash itself, so the
  // position an attacker would have to predict is unpredictable.
  const offset = mac[mac.length - 1]! & 0x0f;
  const binary =
    ((mac[offset]! & 0x7f) << 24) |
    ((mac[offset + 1]! & 0xff) << 16) |
    ((mac[offset + 2]! & 0xff) << 8) |
    (mac[offset + 3]! & 0xff);
  // The 0x7f mask is an interoperability fix, not a security measure: it makes
  // signed and unsigned 32-bit languages agree.

  // Step 4: reduce and zero-pad. One code in ten starts with a zero, and losing
  // it is the classic "it rejects my code sometimes" bug.
  return (binary % 10 ** digits).toString().padStart(digits, '0');
}

/** Computes the code for an instant. */
export async function generateTotp(
  secret: string,
  at: Date = new Date(),
  digits: number = TOTP_DIGITS,
  period: number = TOTP_PERIOD,
  algorithm: TotpAlgorithm = 'SHA1',
): Promise<string> {
  return generateForCounter(secret, counterFor(at, period), digits, algorithm);
}

/**
 * Compares two strings without leaking their contents through timing.
 *
 * A plain `===` on strings can return early at the first differing character,
 * and the timing difference tells an attacker how many leading characters they
 * guessed correctly - which turns a 10^6 search into roughly 60 guesses.
 *
 * This walks every character regardless and accumulates the differences.
 */
export function timingSafeEqual(a: string, b: string): boolean {
  if (a.length !== b.length) return false;

  let difference = 0;
  for (let i = 0; i < a.length; i++) {
    difference |= a.charCodeAt(i) ^ b.charCodeAt(i);
  }

  return difference === 0;
}

export type VerifyOptions = {
  secret: string;
  code: string;
  at?: Date;
  digits?: number;
  period?: number;
  algorithm?: TotpAlgorithm;
  window?: number;
  /**
   * The last counter that already succeeded for this device.
   *
   * Anything at or below it is refused as a replay. Pass it whenever you have
   * it - without this, a code read over a shoulder or captured by a proxy stays
   * usable for the rest of its window.
   */
  lastUsedCounter?: number;
};

export type VerifyResult =
  | { valid: true; counter: number }
  | { valid: false; reason: 'mismatch' | 'replayed' | 'malformed' };

/**
 * Verifies a submitted code.
 *
 * Checks the current step plus `window` steps either side, in constant time,
 * and refuses any counter that has already been used.
 */
export async function verifyTotp(options: VerifyOptions): Promise<VerifyResult> {
  const {
    secret,
    code,
    at = new Date(),
    digits = TOTP_DIGITS,
    period = TOTP_PERIOD,
    algorithm = 'SHA1',
    window = TOTP_DRIFT_WINDOW,
    lastUsedCounter,
  } = options;

  const submitted = code.replace(/\s/g, '');

  if (!new RegExp(`^\\d{${digits}}$`).test(submitted)) {
    return { valid: false, reason: 'malformed' };
  }

  const current = counterFor(at, period);
  let replayed = false;

  for (let offset = -window; offset <= window; offset++) {
    const counter = current + offset;
    if (counter < 0) continue;

    let expected: string;
    try {
      expected = await generateForCounter(secret, counter, digits, algorithm);
    } catch {
      return { valid: false, reason: 'malformed' };
    }

    if (timingSafeEqual(submitted, expected)) {
      // The code is genuine. Whether it is *acceptable* is a separate question:
      // a counter that already succeeded is a replay, not a fresh proof.
      if (lastUsedCounter !== undefined && counter <= lastUsedCounter) {
        replayed = true;
        continue;
      }
      return { valid: true, counter };
    }
  }

  return { valid: false, reason: replayed ? 'replayed' : 'mismatch' };
}

/** Whole seconds until the current code rolls. */
export function secondsRemaining(at: Date = new Date(), period: number = TOTP_PERIOD): number {
  return period - (Math.floor(at.getTime() / 1000) % period);
}

File: authenticator-app/lib/services/totp_service.dart
/// The TOTP engine.
///
/// This is the heart of the whole project, and it is written out by hand rather
/// than delegated to a package. In production you should absolutely use a
/// maintained library (`otp`, `dart_otp`) - `SECURITY.md` says so. Here the
/// point is to open the box: every step below maps to a numbered step in
/// `docs/totp-working.md`, and the whole algorithm is about sixty lines.
///
/// The one-sentence version: a one-time code is a keyed hash of the current
/// half-minute, squeezed down to six digits.
///
/// Nothing in this file performs I/O. No network, no storage, no logging. It
/// takes a secret and a time, and returns digits.
library;

import 'dart:convert';
import 'dart:typed_data';

// Prefixed because this file defines an enum value named `sha1`, which would
// otherwise shadow the hash of the same name.
import 'package:crypto/crypto.dart' as crypto;

/// Hash used inside the HMAC. RFC 6238 permits all three.
///
/// SHA-1 is the default because it is what effectively every deployed
/// authenticator assumes, and interoperability decides this in practice. It is
/// not a weakness here: SHA-1's break is in *collision resistance*, and HMAC
/// does not rely on collision resistance.
enum TotpAlgorithm {
  sha1('SHA1'),
  sha256('SHA256'),
  sha512('SHA512');

  const TotpAlgorithm(this.label);

  /// The spelling used in an `otpauth://` URI.
  final String label;

  static TotpAlgorithm fromLabel(String? value) {
    return switch (value?.toUpperCase().replaceAll('-', '')) {
      'SHA256' => TotpAlgorithm.sha256,
      'SHA512' => TotpAlgorithm.sha512,
      _ => TotpAlgorithm.sha1,
    };
  }

  /// Written as a switch statement rather than a switch expression: the three
  /// arms have different concrete types, and expression inference widens their
  /// common supertype to `Object`. Returning from each arm checks against the
  /// declared `crypto.Hash` instead.
  crypto.Hash get hash {
    switch (this) {
      case TotpAlgorithm.sha1:
        return crypto.sha1;
      case TotpAlgorithm.sha256:
        return crypto.sha256;
      case TotpAlgorithm.sha512:
        return crypto.sha512;
    }
  }
}

/// Thrown when a secret is not valid Base32.
class InvalidSecretException implements Exception {
  const InvalidSecretException(this.message);
  final String message;

  @override
  String toString() => 'InvalidSecretException: $message';
}

/// RFC 4648 Base32, without padding on encode.
///
/// Base32 rather than Base64 for a human reason: the alphabet has no `0`/`O` or
/// `1`/`l`/`I` confusion, so a setup key can be read aloud or typed by hand when
/// a camera is not available.
abstract final class Base32 {
  static const _alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';

  /// Decodes Base32 text into bytes.
  ///
  /// Tolerant on input by design: authenticator users paste keys that arrive
  /// lowercase, in groups of four, or with `=` padding. All of those mean the
  /// same key, and rejecting them would be pedantry that costs a user their
  /// enrollment.
  static Uint8List decode(String input) {
    final cleaned = input
        .toUpperCase()
        .replaceAll(RegExp(r'[\s-]'), '')
        .replaceAll('=', '');

    if (cleaned.isEmpty) {
      throw const InvalidSecretException('The secret is empty.');
    }

    final output = <int>[];
    var buffer = 0;
    var bits = 0;

    for (final char in cleaned.codeUnits) {
      final value = _alphabet.indexOf(String.fromCharCode(char));
      if (value < 0) {
        throw InvalidSecretException(
          'Character "${String.fromCharCode(char)}" is not valid Base32.',
        );
      }

      // Accumulate 5 bits per character, emitting a byte whenever 8 are ready.
      buffer = (buffer << 5) | value;
      bits += 5;

      if (bits >= 8) {
        output.add((buffer >> (bits - 8)) & 0xFF);
        bits -= 8;
      }
    }

    // Any leftover bits are padding, not data, and are discarded.
    if (output.isEmpty) {
      throw const InvalidSecretException(
        'The secret is too short to be a key.',
      );
    }

    return Uint8List.fromList(output);
  }

  /// Encodes bytes as unpadded Base32.
  static String encode(Uint8List bytes) {
    var buffer = 0;
    var bits = 0;
    final output = StringBuffer();

    for (final byte in bytes) {
      buffer = (buffer << 8) | byte;
      bits += 8;

      while (bits >= 5) {
        output.write(_alphabet[(buffer >> (bits - 5)) & 31]);
        bits -= 5;
      }
    }

    if (bits > 0) {
      output.write(_alphabet[(buffer << (5 - bits)) & 31]);
    }

    return output.toString();
  }

  /// True when the text decodes to at least one byte.
  static bool isValid(String input) {
    try {
      decode(input);
      return true;
    } on InvalidSecretException {
      return false;
    }
  }
}

/// Generates time-based one-time passwords.
///
/// All methods are static and pure: the same secret and the same instant always
/// produce the same code, on any device, with no coordination between them.
/// That property is the entire reason this works offline.
abstract final class TotpService {
  /// Seconds a code stays valid. RFC 6238's recommended default.
  static const int defaultPeriod = 30;

  /// Digits shown to the user. Six is universal; eight is permitted.
  static const int defaultDigits = 6;

  /// Computes the code for a point in time.
  ///
  /// [at] defaults to now. Pass it explicitly in tests - never read the clock
  /// inside a test you want to stay green next year.
  static String generate({
    required String secret,
    DateTime? at,
    int digits = defaultDigits,
    int period = defaultPeriod,
    TotpAlgorithm algorithm = TotpAlgorithm.sha1,
  }) {
    final counter = counterFor(at: at, period: period);
    return generateForCounter(
      secret: secret,
      counter: counter,
      digits: digits,
      algorithm: algorithm,
    );
  }

  /// The time step: how many whole [period]s have elapsed since the Unix epoch.
  ///
  /// This is the only value the phone and the server share at verification time.
  /// UTC is not optional - `millisecondsSinceEpoch` is already UTC-based, which
  /// is exactly why local time never enters this calculation. Using local time
  /// here is the classic bug: it works on your laptop and fails for every user
  /// in another timezone.
  static int counterFor({DateTime? at, int period = defaultPeriod}) {
    final instant = at ?? DateTime.now();
    final seconds = instant.millisecondsSinceEpoch ~/ 1000;
    return seconds ~/ period;
  }

  /// HOTP: the counter-based code TOTP is built on (RFC 4226).
  static String generateForCounter({
    required String secret,
    required int counter,
    int digits = defaultDigits,
    TotpAlgorithm algorithm = TotpAlgorithm.sha1,
  }) {
    final key = Base32.decode(secret);

    // Step 1: the counter as an 8-byte big-endian integer.
    final message = Uint8List(8);
    var remaining = counter;
    for (var i = 7; i >= 0; i--) {
      message[i] = remaining & 0xFF;
      remaining >>= 8;
    }

    // Step 2: HMAC. A keyed hash - unforgeable without the key, and it leaks
    // nothing about the key. An attacker who watches a thousand codes still
    // cannot derive the secret.
    final mac = crypto.Hmac(algorithm.hash, key).convert(message).bytes;

    // Step 3: dynamic truncation. Which four bytes to take is derived from the
    // hash itself, so the position is unpredictable. Always taking the first
    // four would be.
    final offset = mac[mac.length - 1] & 0x0F;
    final binary =
        ((mac[offset] & 0x7F) << 24) |
        ((mac[offset + 1] & 0xFF) << 16) |
        ((mac[offset + 2] & 0xFF) << 8) |
        (mac[offset + 3] & 0xFF);
    // The 0x7F mask clears the top bit. That is an interoperability fix, not a
    // security measure: it makes languages with signed 32-bit integers agree
    // with languages that have unsigned ones.

    // Step 4: reduce to N digits, and pad. One code in ten starts with a zero,
    // and dropping it is a bug users experience as "it rejects my code
    // sometimes".
    final modulo = _pow10(digits);
    return (binary % modulo).toString().padLeft(digits, '0');
  }

  /// Whole seconds until the current code expires.
  ///
  /// Counts down from [period] to 1. The user is told "expires in 24 seconds",
  /// so 0 would be a lie - at zero the code has already rolled.
  static int secondsRemaining({DateTime? at, int period = defaultPeriod}) {
    final instant = at ?? DateTime.now();
    final seconds = instant.millisecondsSinceEpoch ~/ 1000;
    return period - (seconds % period);
  }

  /// Progress through the current step, from 1.0 (fresh) down to 0.0 (expiring).
  ///
  /// Uses milliseconds so the ring animates smoothly rather than ticking in
  /// one-second jumps.
  static double progress({DateTime? at, int period = defaultPeriod}) {
    final instant = at ?? DateTime.now();
    final millis = instant.millisecondsSinceEpoch;
    final periodMillis = period * 1000;
    final elapsed = millis % periodMillis;
    return 1.0 - (elapsed / periodMillis);
  }

  /// Formats a code for display: `482931` becomes `482 931`.
  ///
  /// Grouping halves the chance of a transcription slip when someone is copying
  /// six digits by eye.
  static String formatForDisplay(String code) {
    if (code.length != 6) return code;
    return '${code.substring(0, 3)} ${code.substring(3)}';
  }

  static int _pow10(int exponent) {
    var result = 1;
    for (var i = 0; i < exponent; i++) {
      result *= 10;
    }
    return result;
  }
}

/// Helper for tests and for building a secret from raw bytes.
Uint8List asciiBytes(String text) => Uint8List.fromList(ascii.encode(text));
